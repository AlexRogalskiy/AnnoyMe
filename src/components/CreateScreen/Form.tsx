import React, { useState } from "react";
import { View, TextInput } from "react-native";
import { TaskSchedule, TaskChanges } from "../../types";
import ScheduleInput from "../ScheduleInput";
import styles from "./styles";

interface Props {
  title: string;
  schedule: TaskSchedule;

  startHour: number;
  endHour: number;

  onChange: (changes: TaskChanges) => void;
  onValidationChange: (isValid: boolean) => void;
}

const validateTitle = (title: string) => title.length > 0;

export default ({
  title,
  schedule,
  startHour,
  endHour,
  onChange,
  onValidationChange,
}: Props) => {
  const [formTitle, setFormTitle] = useState(title);
  const [isTitleValid, setIsTitleValid] = useState(validateTitle(formTitle));
  const [formSchedule, setFormSchedule] = useState(schedule);

  const onTitleChange = (newTitle: string) => {
    setFormTitle(newTitle);
    const isValid = validateTitle(newTitle);
    if (isValid !== isTitleValid) {
      setIsTitleValid(isValid);
      onValidationChange(isValid);
    }

    onChange({ title: newTitle });
  };

  const onScheduleChange = (newSchedule: TaskSchedule) => {
    setFormSchedule(newSchedule);

    onChange({ schedule: newSchedule });
  };

  return (
    <View style={styles.container}>
      <TextInput
        style={[
          styles.input,
          isTitleValid ? styles.inputValid : styles.inputInvalid,
        ]}
        placeholder="Name of the annoyance"
        onChangeText={onTitleChange}
        defaultValue={formTitle}
      />
      <ScheduleInput
        schedule={formSchedule}
        startHour={startHour}
        endHour={endHour}
        onChange={onScheduleChange}
      />
    </View>
  );
};
