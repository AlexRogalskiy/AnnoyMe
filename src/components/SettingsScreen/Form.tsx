import React from "react";
import { View } from "react-native";
import * as constants from "../../constants";
import NumberSettingsInput from "./NumberSettingsInput";
import styles from "./styles";

const isValidHour = (hour: number) => hour >= 0 && hour <= 24;

const isValidFrequency = (minutes: number) =>
  minutes >= constants.MIN_FREQUENCY && minutes <= 120;

interface Props {
  startHour: number;
  endHour: number;
  frequency: number;

  onNumberSettingsChange: (key: string, value: number) => void;
}

export default ({
  startHour,
  endHour,
  frequency,
  onNumberSettingsChange,
}: Props) => (
  <View style={styles.container}>
    <NumberSettingsInput
      label="Annoy from hour"
      value={startHour}
      validate={isValidHour}
      onValidChange={val => onNumberSettingsChange("startHour", val)}
    />
    <NumberSettingsInput
      label="Annoy till hour"
      value={endHour}
      validate={isValidHour}
      onValidChange={val => onNumberSettingsChange("endHour", val)}
    />
    <NumberSettingsInput
      label="Annoys frequency"
      value={frequency}
      validate={isValidFrequency}
      onValidChange={val => onNumberSettingsChange("frequency", val)}
    />
  </View>
);
