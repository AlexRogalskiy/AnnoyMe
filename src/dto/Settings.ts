export default interface Settings {
  startHour: number;
  endHour: number;
  frequency: number;
}
