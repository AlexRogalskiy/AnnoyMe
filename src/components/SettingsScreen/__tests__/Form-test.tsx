import React from "react";
import renderer, { act } from "react-test-renderer";
import Settings from "../../../dto/Settings";
import Form from "../Form";

const settings = {
  startHour: 10,
  endHour: 23,
  frequency: 45,
};

const onChangeIgnore = (_: Settings) => undefined;

it("Form renders settings form", () => {
  const tree = renderer
    .create(<Form settings={settings} onChange={onChangeIgnore} />)
    .toJSON();

  expect(tree).toMatchSnapshot("Form-settings");
});

const assertInputChange = (
  testID: string,
  value: number,
  settingsBefore: Settings,
  settingsAfter: Settings,
) => {
  let changeableSettings: Settings | undefined;
  const onChange = (newSettings: Settings) =>
    (changeableSettings = newSettings);

  const component = renderer.create(
    <Form settings={settingsBefore} onChange={onChange} />,
  ).root;

  const input = component.findByProps({ testID });
  expect(input).toBeTruthy();

  act(() => {
    input.props.onChange(value);
  });
  expect(changeableSettings).toEqual(settingsAfter);
};

it("Form changes start hour", () =>
  assertInputChange("start-hour-input", 14, settings, {
    startHour: 14,
    endHour: 23,
    frequency: 45,
  }));

it("Form changes end hour", () =>
  assertInputChange("end-hour-input", 21, settings, {
    startHour: 10,
    endHour: 21,
    frequency: 45,
  }));

it("Form changes frequency", () =>
  assertInputChange("frequency-input", 100, settings, {
    startHour: 10,
    endHour: 23,
    frequency: 100,
  }));
