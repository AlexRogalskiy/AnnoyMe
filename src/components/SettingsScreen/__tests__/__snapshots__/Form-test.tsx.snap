// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form renders settings form: Form-settings 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "justifyContent": "flex-start",
      "padding": 20,
    }
  }
>
  <View
    style={
      Object {
        "paddingBottom": 10,
      }
    }
  >
    <Text
      style={
        Object {
          "fontWeight": "200",
        }
      }
    >
      Annoy from hour
    </Text>
    <TextInput
      allowFontScaling={true}
      defaultValue="10"
      keyboardType="number-pad"
      onChangeText={[Function]}
      rejectResponderTermination={true}
      style={
        Array [
          Object {
            "borderBottomWidth": 1,
          },
          Object {
            "borderBottomColor": "#c8e6c9",
          },
        ]
      }
      testID="number-settings-input-text-input"
      underlineColorAndroid="transparent"
    />
  </View>
  <View
    style={
      Object {
        "paddingBottom": 10,
      }
    }
  >
    <Text
      style={
        Object {
          "fontWeight": "200",
        }
      }
    >
      Annoy till hour
    </Text>
    <TextInput
      allowFontScaling={true}
      defaultValue="23"
      keyboardType="number-pad"
      onChangeText={[Function]}
      rejectResponderTermination={true}
      style={
        Array [
          Object {
            "borderBottomWidth": 1,
          },
          Object {
            "borderBottomColor": "#c8e6c9",
          },
        ]
      }
      testID="number-settings-input-text-input"
      underlineColorAndroid="transparent"
    />
  </View>
  <View
    style={
      Object {
        "paddingBottom": 10,
      }
    }
  >
    <Text
      style={
        Object {
          "fontWeight": "200",
        }
      }
    >
      Annoys frequency
    </Text>
    <TextInput
      allowFontScaling={true}
      defaultValue="45"
      keyboardType="number-pad"
      onChangeText={[Function]}
      rejectResponderTermination={true}
      style={
        Array [
          Object {
            "borderBottomWidth": 1,
          },
          Object {
            "borderBottomColor": "#c8e6c9",
          },
        ]
      }
      testID="number-settings-input-text-input"
      underlineColorAndroid="transparent"
    />
  </View>
</View>
`;
